File: frontend/src/lib/faceCalibrationConsent.ts
/**
 * Server-side face/delivery calibration consent (mirror of
 * `deliveryAnalyticsConsent.ts`).
 *
 * The calibration *profile* stays on the device, but consent for it is recorded
 * on the user row so it is demonstrable (GDPR Art. 7(1) / BIPA) and per-user
 * (no leak across accounts that share a browser). `hasActiveFaceCalibrationConsent`
 * encodes the version gate: a stored consent at an older notice version is treated
 * as not-active, which re-prompts on `/calibrate` and triggers clearing the stale
 * local baseline.
 *
 * FACE_CALIBRATION_NOTICE_VERSION mirrors the backend constant in
 * `app/services/face_calibration_consent.py` — bump BOTH when the notice copy
 * changes.
 */

import type { MeResponse } from '../types/user';

export const FACE_CALIBRATION_NOTICE_VERSION = 1 as const;

export function hasActiveFaceCalibrationConsent(
  me: Pick<
    MeResponse,
    | 'face_calibration_consent_at'
    | 'face_calibration_consent_version'
    | 'face_calibration_revoked_at'
  > | null,
): boolean {
  if (!me?.face_calibration_consent_at) return false;
  if (me.face_calibration_consent_version !== FACE_CALIBRATION_NOTICE_VERSION) {
    return false;
  }
  if (!me.face_calibration_revoked_at) return true;

  const consentAt = Date.parse(me.face_calibration_consent_at);
  const revokedAt = Date.parse(me.face_calibration_revoked_at);
  if (Number.isNaN(consentAt)) return false;
  if (Number.isNaN(revokedAt)) return true;
  return revokedAt < consentAt;
}

File: backend/app/services/face_calibration_consent.py
"""Consent helpers for browser-local face/delivery calibration.

The calibration profile (a small set of aggregate face-geometry baselines used
by the webcam delivery heuristic) stays on the candidate's device and never
reaches our server. The *consent* to perform calibration, however, lives on the
user row so it is demonstrable (GDPR Art. 7(1) / BIPA) and cannot leak across
accounts that share a browser.

Unlike `delivery_consent.py`, there is NO server-side artifact to purge on
revocation — the only stored data is the local profile, which the client clears
itself. So this module is just the version constant + the active-consent check;
the revoke endpoint simply stamps `face_calibration_revoked_at`.
"""

from app.db.models.user import User


# Notice version. Bump whenever the calibration privacy notice copy changes
# (mirror the bump in the frontend's faceCalibrationConsent.ts). A stored
# consent at an older version is treated as not-active, so `/calibrate`
# re-prompts and the stale local baseline is cleared. Deliberately decoupled
# from CURRENT_BIOMETRIC_VERSION (policy_versions.py), mirroring how
# DELIVERY_ANALYTICS_NOTICE_VERSION is its own constant.
FACE_CALIBRATION_NOTICE_VERSION = 1


def has_active_face_calibration_consent(user: User) -> bool:
    """Return True only for the current, non-revoked notice version."""
    if user.face_calibration_consent_at is None:
        return False
    if user.face_calibration_consent_version != FACE_CALIBRATION_NOTICE_VERSION:
        return False
    revoked_at = user.face_calibration_revoked_at
    return revoked_at is None or revoked_at < user.face_calibration_consent_at
